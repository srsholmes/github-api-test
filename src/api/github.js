import axios from 'axios';

const githubGraphQL = axios.create({
  baseURL: 'https://api.github.com/graphql',
  headers: {
    Authorization: `bearer ${process.env.GITHUB_PERSONAL_ACCESS_TOKEN}`
  }
});

const SEARCH_USERS = `
  query ($user: String!) {
    search(query: $user, type: USER, first: 10) {
      edges {
        node {
          ... on User {
            avatarUrl,
            name,
            login
          }
        }
      }
    }
  }
`;

const ACTIVITY = `
  query ($user: String!) {
    repositoryOwner(login: $user) {
      ... on User {
        login
        avatarUrl
        contributionsCollection {
          startedAt
          endedAt
          commitContributionsByRepository(maxRepositories:10) {
            repository {
              name,
            }
            contributions(first:5) {
              edges {
                node {
                  occurredAt
                }
              }
            }
          }
        }
      }
    }
  }
`;

export const searchUser = async val => {
  const res = await githubGraphQL.post('', {
    query: SEARCH_USERS,
    variables: {user: val}
  });
  console.log({res});
  return res.data.data.search.edges;
};

export const getUserActivity = async val => {
  const res = await githubGraphQL.post('', {
    query: ACTIVITY,
    variables: {user: val}
  });
  return res.data.data.repositoryOwner.contributionsCollection
    .commitContributionsByRepository;
};
