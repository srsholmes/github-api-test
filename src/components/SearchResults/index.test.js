import React from 'react';
import {render, getAllByTestId, fireEvent} from '@testing-library/react';
import {SearchResults} from './index';

import * as dependency from '../../api/github';

jest.mock('../../api/github', () => ({
  getUserActivity: jest.fn(() => Promise.resolve([]))
}));

describe('SearchResults', () => {
  const MOCK_RESULTS = [
    {
      node: {
        avatarUrl: 'https://avatars3.githubusercontent.com/u/39902?v=4',
        name: 'Ben Straub',
        login: 'ben'
      }
    },
    {
      node: {
        avatarUrl: 'https://avatars3.githubusercontent.com/u/282759?v=4',
        name: 'Ben Balter',
        login: 'benbalter'
      }
    },
    {
      node: {
        avatarUrl: 'https://avatars0.githubusercontent.com/u/5750?v=4',
        name: 'Ben Newman',
        login: 'benjamn'
      }
    }
  ];

  it('displays the message when there are no results', () => {
    const {container} = render(<SearchResults results={[]} />);
    expect(container).toMatchSnapshot();
  });

  it('Results have the Get User Activity button', () => {
    const {container} = render(<SearchResults results={MOCK_RESULTS} />);
    const [button] = getAllByTestId(container, 'test-get-activity');
    expect(button.textContent).toBe('Get User Activity');
    fireEvent.click(button);
    expect(dependency.getUserActivity).toHaveBeenCalledTimes(1);
    expect(dependency.getUserActivity).toBeCalledWith('ben');
  });

  it('displays results when there are results', () => {
    const {container} = render(<SearchResults results={MOCK_RESULTS} />);
    expect(container).toMatchSnapshot();
  });
});
