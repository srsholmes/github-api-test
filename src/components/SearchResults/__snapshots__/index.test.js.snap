// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults displays results when there are results 1`] = `
<div>
  <ul
    class="wrapper"
    data-testid="user-results"
  >
    <li
      class="result"
    >
      <img
        alt="Ben Straub"
        class="avatar"
        src="https://avatars3.githubusercontent.com/u/39902?v=4"
      />
      <div
        class="infoWrapper"
      >
        <div
          class="info"
        >
          <span
            class="title"
          >
            Name:
          </span>
          <span>
            Ben Straub
          </span>
        </div>
        <div
          class="info"
        >
          <span
            class="title"
          >
            Username:
          </span>
          <span>
            ben
          </span>
        </div>
        <button
          data-testid="test-get-activity"
        >
          Get User Activity
        </button>
      </div>
    </li>
    <li
      class="result"
    >
      <img
        alt="Ben Balter"
        class="avatar"
        src="https://avatars3.githubusercontent.com/u/282759?v=4"
      />
      <div
        class="infoWrapper"
      >
        <div
          class="info"
        >
          <span
            class="title"
          >
            Name:
          </span>
          <span>
            Ben Balter
          </span>
        </div>
        <div
          class="info"
        >
          <span
            class="title"
          >
            Username:
          </span>
          <span>
            benbalter
          </span>
        </div>
        <button
          data-testid="test-get-activity"
        >
          Get User Activity
        </button>
      </div>
    </li>
    <li
      class="result"
    >
      <img
        alt="Ben Newman"
        class="avatar"
        src="https://avatars0.githubusercontent.com/u/5750?v=4"
      />
      <div
        class="infoWrapper"
      >
        <div
          class="info"
        >
          <span
            class="title"
          >
            Name:
          </span>
          <span>
            Ben Newman
          </span>
        </div>
        <div
          class="info"
        >
          <span
            class="title"
          >
            Username:
          </span>
          <span>
            benjamn
          </span>
        </div>
        <button
          data-testid="test-get-activity"
        >
          Get User Activity
        </button>
      </div>
    </li>
  </ul>
</div>
`;

exports[`SearchResults displays the message when there are no results 1`] = `
<div>
  <ul
    class="wrapper"
    data-testid="user-results"
  >
    <p>
      No results found, please try a different search term
    </p>
  </ul>
</div>
`;
